mmaction/datasets/pipelines/augmentations.py

```diff
@@ -868,6 +870,406 @@ def __repr__(self):
+
+            results['imgs'] = outs
+            results['img_shape'] = outs[0].shape[:2]
+            results['gt_bboxes'] = gt_bboxes
+
+        return results
+
+    def __repr__(self):
+        repr_str = (f'{self.__class__.__name__}('
+                    f'expand_ratio={self.expand_ratio}, '
+                    f'max_expand_ratio={self.max_expand_ratio}, '
+                    f'mean_values={self.mean_values})')
+        return repr_str
+
+
+@PIPELINES.register_module()
+class TubeResize:
+    """Resize images and bounding boxes to a specific size.
+
+    Required keys are "gt_bboxes", "img_shape", "imgs", added or modified keys
+    are "imgs", "img_shape", "gt_bboxes", "box_output_shape".
+
+    Args:
+        resize_scale (int | tuple[int]): (w, h) for target shape.
+        output_stride (int): Output stride used for scaling target boxes shape.
+    """
+
+    def __init__(self, resize_scale, output_stride=4):
+        resize_scale = _pair(resize_scale)
+        if not mmcv.is_tuple_of(resize_scale, int):
+            raise TypeError('resize_scale must be int or tuple of int, '
+                            f'but got {resize_scale}')
+
+        self.resize_h, self.resize_w = resize_scale
+
+        self.output_stride = output_stride
+        self.output_h = self.resize_h // self.output_stride
+        self.output_w = self.resize_w // self.output_stride
+
+    def __call__(self, results):
+        gt_bboxes = results['gt_bboxes']
+        origin_h, origin_w = results['img_shape']
+        imgs = results['imgs']
+
+        for label_index in gt_bboxes:
+            for tube in gt_bboxes[label_index]:
+                tube[:, 0::2] = tube[:, 0::2] / origin_w * self.output_w
+                tube[:, 1::2] = tube[:, 1::2] / origin_h * self.output_h
+
+        imgs = [
+            mmcv.imresize(img, (self.resize_w, self.resize_h)) for img in imgs
+        ]
+
+        results['imgs'] = imgs
+        results['img_shape'] = imgs[0].shape[:2]
+        results['gt_bboxes'] = gt_bboxes
+        results['box_output_shape'] = (self.output_h, self.output_w)
+
+        return results
+
+    def __repr__(self):
+        repr = (f'{self.__class__.__name__}('
+                f'resize_scale={(self.resize_h, self.resize_w)}, '
+                f'output_stride={self.output_stride})')
+        return repr
+
+
+@PIPELINES.register_module()
+class MOCTubeExtract:
+    """Extract required features for MOC.
+
+    Extract required features for MOC, including gaussian heatmap, gt_bboxes'
+    height and width feature, gt_bboxes' movement feature, center position for
+    key frame and all frames.
+
+    Args:
+        max_objs (int): Number of max objects. Default: 128.
+    """
+
+    def __init__(self, max_objs=128):
+        self.max_objs = max_objs
+
+    @staticmethod
+    def gaussian_radius(det_size, min_overlap=0.7):
+
+        def solve(a, b, c):
+            return (np.sqrt(b**2 - 4 * a * c) + b) / 2
```

Comment:
what does this formula solve? double check the denominator

Reply:
it is used for generating gaussian featmap
![image](https://user-images.githubusercontent.com/33949370/100619636-6afc7280-3358-11eb-92da-e09f6485fba4.png)